From 0169fd67d86e933d95c0877f56e0de7cfc17ec9e Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Wed, 24 Jul 2019 16:11:37 +0800
Subject: [PATCH] usb: dwc_otg: fix local_fiq_disable with irq disable without
 waiting

In ARM64, the FIQ mechanism used by this driver is not current
implemented, however there are much codes depended on this due to
the ARM32 implemented it. A workaround has been made, please refer to
the commit 2babc87ab (ARM64/DWC_OTG: Port dwc_otg driver to ARM64).

Based on the workaround, there will be call trace as below:

BUG: sleeping function called from invalid context at kernel/irq/manage.c:112
in_atomic(): 1, irqs_disabled(): 128, pid: 0, name: swapper/3
Preemption disabled at:
[<ffffff80080956a4>] secondary_start_kernel+0xbc/0x150
CPU: 3 PID: 0 Comm: swapper/3 Tainted: G        WC        4.18.40-yocto-standard #1
Hardware name: Raspberry Pi 3 Model B+ (DT)
Call trace:
 dump_backtrace+0x0/0x170
 show_stack+0x24/0x30
 dump_stack+0x80/0xa4
 ___might_sleep+0x14c/0x170
 __might_sleep+0x58/0x90
 synchronize_irq+0x54/0xc8
 disable_irq+0x38/0x48
 local_fiq_disable+0x20/0x28
 dwc_otg_hcd_handle_intr+0x2f0/0x3a8
 dwc_otg_hcd_irq+0x20/0x38
 usb_hcd_irq+0x3c/0x60
 __handle_irq_event_percpu+0x78/0x298
 handle_irq_event_percpu+0x40/0x98
 handle_irq_event+0x50/0x80
 handle_level_irq+0xd8/0x160
 generic_handle_irq+0x34/0x50
 bcm2836_chained_handle_irq+0x4c/0x68
 generic_handle_irq+0x34/0x50
 __handle_domain_irq+0x98/0x108
 bcm2836_arm_irqchip_handle_irq+0x68/0xd0
 el1_irq+0xb0/0x128
 arch_cpu_idle+0x38/0x1a8
 do_idle+0x1dc/0x238
 cpu_startup_entry+0x2c/0x30
 secondary_start_kernel+0x140/0x150

In order to avoid the call trace, replace disable_irq with
disable_irq_nosync, this is also a workaround.

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c b/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
index ed9694058ed0..1bb34eca60ff 100644
--- a/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
+++ b/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
@@ -417,7 +417,7 @@ void local_fiq_enable(void)
 void local_fiq_disable(void)
 {
 	if (simfiq_irq >= 0)
-		disable_irq(simfiq_irq);
+		disable_irq_nosync(simfiq_irq);
 }
 
 irqreturn_t fiq_irq_handler(int irq, void *dev_id)
-- 
2.17.1

